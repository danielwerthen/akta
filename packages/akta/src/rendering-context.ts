import { createSyncContext } from './sync-context';
import { BehaviorSubject, Subscription } from 'rxjs';

export enum RenderingState {
  init = 'init',
  active = 'active',
  terminated = 'terminated',
}

export type Detachable = ChildNode & {
  onDetach?: () => void;
};

export class RenderingContext {
  subscriptions: Subscription[] = [];
  teardowns: Function[] = [];
  dependencies: number = 0;
  state: BehaviorSubject<RenderingState> = new BehaviorSubject(
    RenderingState.init
  );
  spawn() {
    const child = new RenderingContext();
    this.teardowns.push(() => child.terminate());
    return child;
  }
  terminate() {
    for (let sub of this.subscriptions) {
      sub.unsubscribe();
    }
    for (let cb of this.teardowns) {
      cb();
    }
    this.state.next(RenderingState.terminated);
  }
  placeholder(): ChildNode {
    const el: Detachable = document.createTextNode('');
    el.onDetach = () => {
      this.dependencies -= 1;
      if (this.dependencies === 0) {
        this.state.next(RenderingState.active);
      }
    };
    this.dependencies += 1;
    return el;
  }
  monitor(): void {
    if (this.dependencies < 1) {
      this.state.next(RenderingState.active);
      return;
    }
  }
}

export const renderingContext = createSyncContext<RenderingContext>();
