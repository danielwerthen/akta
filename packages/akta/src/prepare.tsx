import { AktaNode, isAktaElement } from './types';
import { isObservable } from 'rxjs';
import { RenderingContext, RenderingState } from './rendering-context';
import { renderElement } from './render-element';
import { DependencyMap } from './dependency-map';

export type NodeInstance = ChildNode | [ChildNode, ...ChildNode[]];

export type Detachable = ChildNode & {
  onDetach?: () => void;
};

function teardown(item: ChildNode) {
  const detachable = item as Detachable;
  if (typeof detachable.onDetach === 'function') {
    detachable.onDetach();
  }
}

function remove(item: NodeInstance) {
  if (Array.isArray(item)) {
    item.forEach(remove);
  } else {
    item.remove();
    teardown(item);
  }
}

function replace(next: NodeInstance, old: NodeInstance) {
  let anchor: undefined | ChildNode = undefined;
  if (Array.isArray(old)) {
    for (const item of old) {
      if (typeof anchor === 'undefined') {
        anchor = item as ChildNode;
      } else {
        remove(item);
      }
    }
  } else {
    anchor = old;
  }
  if (!anchor) {
    throw new Error('Node structure is malformed');
  }
  if (Array.isArray(next)) {
    anchor.replaceWith(...next);
    teardown(anchor);
  } else {
    anchor.replaceWith(next);
    teardown(anchor);
  }
}

export function prepare(
  blueprint: AktaNode,
  ctx: RenderingContext,
  deps: DependencyMap,
  recursive: boolean = false
): NodeInstance {
  try {
    if (Array.isArray(blueprint)) {
      const result: Array<ChildNode> = [];
      for (let item of blueprint) {
        const prep = prepare(item, ctx, deps, true);
        if (Array.isArray(prep)) {
          result.push(...prep);
        } else {
          result.push(prep);
        }
      }
      if (result.length === 0) {
        result.push(document.createTextNode(''));
      }
      return result as NodeInstance;
    } else if (isObservable(blueprint)) {
      let instance: NodeInstance | null = null;
      let activeSpawn: RenderingContext | null = null;
      const sub = blueprint.subscribe({
        next: next => {
          if (activeSpawn) {
            activeSpawn.terminate();
          }
          activeSpawn = ctx.spawn();
          const nextInstance = prepare(next, activeSpawn, deps);
          if (activeSpawn.state.value === RenderingState.init) {
            ctx.dependencies += 1;
            const sub = activeSpawn.state.subscribe(state => {
              if (state !== RenderingState.init) {
                sub.unsubscribe();
              }
            });
            sub.add(() => {
              ctx.dependencies -= 1;
            });
            ctx.subscriptions.push(sub);
          }
          if (instance) {
            replace(nextInstance, instance);
          }
          instance = nextInstance;
        },
      });
      ctx.teardowns.push(() => activeSpawn?.terminate());
      ctx.subscriptions.push(sub);
      if (!instance) {
        instance = ctx.placeholder();
      }
      return instance;
    } else if (blueprint === null) {
      const empty = document.createTextNode('');
      return empty;
    } else if (typeof blueprint === 'string') {
      return document.createTextNode(blueprint);
    } else if (typeof blueprint === 'number') {
      return document.createTextNode(blueprint.toString());
    } else if (isAktaElement(blueprint)) {
      return renderElement(blueprint, ctx, deps);
    } else {
      return blueprint as HTMLElement;
    }
  } finally {
    if (!recursive) {
      ctx.monitor();
    }
  }
}
