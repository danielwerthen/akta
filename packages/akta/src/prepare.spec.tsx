/** @jsxImportSource . */
import { AktaNode, DependencyMap } from './index';
import { from, Observable, of, Subject } from 'rxjs';
import { prepare } from './prepare';
import { RenderingContext, RenderingState } from './rendering-context';

function render(
  blueprint: AktaNode,
  type: string = 'div',
  ctx: RenderingContext = new RenderingContext()
) {
  const parent = document.createElement(type);
  const node = prepare(blueprint, ctx, new DependencyMap());
  if (Array.isArray(node)) {
    parent.append(...node);
  } else {
    parent.append(node);
  }
  return parent;
}

describe('prepare', () => {
  it('takes nested arrays of elements', () => {
    const el = render([
      [render('alpha', 'p'), [render('beta', 'p')]],
      render('zeta', 'p'),
    ]);
    expect(el).toMatchSnapshot();
  });

  it('takes observable nodes and inserts them in the correct order', () => {
    const sub1 = new Subject<AktaNode>();
    const sub2 = new Subject<AktaNode>();
    const sub3 = new Subject<AktaNode>();
    const ctx = new RenderingContext();
    const el = render(
      [of(document.createTextNode('alpha')), sub1, sub2, sub3],
      'div',
      ctx
    );
    expect(ctx.state.value).toEqual(RenderingState.init);
    sub3.next(render('sub3', 'p'));
    expect(el).toMatchSnapshot('sub3');
    expect(ctx.state.value).toEqual(RenderingState.init);
    sub1.next(render('sub1', 'p'));
    expect(el).toMatchSnapshot('sub1');
    expect(ctx.state.value).toEqual(RenderingState.init);
    sub2.next(render('sub2', 'p'));
    expect(el).toMatchSnapshot('sub2');
    expect(ctx.state.value).toEqual(RenderingState.active);
    sub2.next(null);
    expect(el).toMatchSnapshot('sub2-empty');
    sub1.next(render('sub1', 'span'));
    expect(el).toMatchSnapshot('sub1-empty');
    sub3.next(null);
    expect(el).toMatchSnapshot('sub3-empty');
  });

  it('unsubscribes to removed elements', () => {
    const sub = new Subject<AktaNode>();
    const onUnSub = jest.fn();
    const obs = new Observable<AktaNode>(sub => {
      sub.next(document.createTextNode('Observable!'));
      return onUnSub;
    });
    const ctx = new RenderingContext();
    const el = render(sub, 'div', ctx);
    expect(el).toMatchSnapshot('initial');
    expect(ctx.state.value).toEqual(RenderingState.init);
    sub.next([render('alpha', 'p'), obs]);
    expect(ctx.state.value).toEqual(RenderingState.active);
    expect(el).toMatchSnapshot('with obs');
    expect(onUnSub.mock.calls.length).toEqual(0);
    sub.next([render('alpha', 'p')]);
    expect(el).toMatchSnapshot('without obs');
    expect(onUnSub.mock.calls.length).toEqual(1);
  });
  it('takes nested arrays of observables of arrays', () => {
    const el = render([
      from([
        [render('alpha'), render('beta')],
        [render('daniel'), render('kalle')],
      ]),
      render('zeta', 'p'),
    ]);
    expect(el).toMatchSnapshot();
  });
  it('handles jsx', () => {
    const el = render([<p className="daniel">Akta {of(5)}</p>]);
    expect(el).toMatchSnapshot();
  });
});
