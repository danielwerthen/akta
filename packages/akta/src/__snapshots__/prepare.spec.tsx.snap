// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepare handles async generator components 1`] = `
<div>
  <p>
    Component 
    42
  </p>
</div>
`;

exports[`prepare handles components 1`] = `
<div>
  <p>
    Component
  </p>
</div>
`;

exports[`prepare handles generator components 1`] = `
<div>
  <p>
    Initial
  </p>
</div>
`;

exports[`prepare handles jsx 1`] = `
<div>
  <p
    class="daniel"
  >
    Akta 
    <span>
      Number 5
    </span>
  </p>
</div>
`;

exports[`prepare takes nested arrays of elements 1`] = `
<div>
  <p>
    alpha
  </p>
  <p>
    beta
  </p>
  <p>
    zeta
  </p>
</div>
`;

exports[`prepare takes nested arrays of observables of arrays 1`] = `
<div>
  <div>
    daniel
  </div>
  <div>
    kalle
  </div>
  <p>
    zeta
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub1 1`] = `
<div>
  alpha
  <p>
    sub1
  </p>
  
  <p>
    sub3
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub1-empty 1`] = `
<div>
  alpha
  <span>
    sub1
  </span>
  
  <p>
    sub3
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub2 1`] = `
<div>
  alpha
  <p>
    sub1
  </p>
  <p>
    sub2
  </p>
  <p>
    sub3
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub2-empty 1`] = `
<div>
  alpha
  <p>
    sub1
  </p>
  
  <p>
    sub3
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub3 1`] = `
<div>
  alpha
  
  
  <p>
    sub3
  </p>
</div>
`;

exports[`prepare takes observable nodes and inserts them in the correct order: sub3-empty 1`] = `
<div>
  alpha
  <span>
    sub1
  </span>
  
  
</div>
`;

exports[`prepare unsubscribes to removed elements: initial 1`] = `
<div>
  
</div>
`;

exports[`prepare unsubscribes to removed elements: with obs 1`] = `
<div>
  <p>
    alpha
  </p>
  Observable!
</div>
`;

exports[`prepare unsubscribes to removed elements: without obs 1`] = `
<div>
  <p>
    alpha
  </p>
</div>
`;
